File: artifacts/api-server/src/routes/challenges.ts
// Rutas de retos: retos predefinidos y retos personalizados por docentes
// Los retos son el corazon de la plataforma de competencia
import { Router } from "express";
import { db } from "@workspace/db";
import {
  retosTable,
  retosPersonalizadosTable,
  retosPersonalizadosPreguntasTable,
  preguntasTable,
  modulosTable,
  nivelesTable,
  perfilesTable,
} from "@workspace/db";
import { eq, and, sql } from "drizzle-orm";
import { requireAuth, requireDocente, type AuthRequest } from "../lib/auth";

const router = Router();

// GET /challenges - Obtener todos los retos predefinidos activos (soporta filtro ?id_modulo=N)
router.get("/challenges", requireAuth, async (req, res) => {
  try {
    const idModulo = req.query["id_modulo"] ? Number(req.query["id_modulo"]) : null;

    const retos = await db
      .select()
      .from(retosTable)
      .where(
        idModulo
          ? and(eq(retosTable.activo, true), eq(retosTable.id_modulo, idModulo))
          : eq(retosTable.activo, true)
      );

    // Agrega informacion de modulo y nivel a cada reto
    const retosConInfo = await Promise.all(
      retos.map(async (reto) => {
        const [modulo] = await db.select().from(modulosTable).where(eq(modulosTable.id, reto.id_modulo));
        const [nivel] = await db.select().from(nivelesTable).where(eq(nivelesTable.id, reto.id_nivel));
        return { ...reto, modulo: modulo || null, nivel: nivel || null };
      })
    );

    res.json(retosConInfo);
  } catch (err) {
    req.log.error({ err }, "Error al obtener retos");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /challenges/:challengeId - Obtener un reto con sus preguntas
router.get("/challenges/:challengeId", requireAuth, async (req, res) => {
  try {
    const challengeId = Number(req.params.challengeId);
    const [reto] = await db.select().from(retosTable).where(eq(retosTable.id, challengeId));

    if (!reto) {
      res.status(404).json({ error: "not_found", message: "Reto no encontrado" });
      return;
    }

    // Obtiene las preguntas del nivel y modulo correspondiente
    const preguntas = await db
      .select()
      .from(preguntasTable)
      .where(
        and(
          eq(preguntasTable.id_modulo, reto.id_modulo),
          eq(preguntasTable.id_nivel, reto.id_nivel),
          eq(preguntasTable.activa, true)
        )
      )
      .limit(reto.numero_preguntas);

    const [modulo] = await db.select().from(modulosTable).where(eq(modulosTable.id, reto.id_modulo));
    const [nivel] = await db.select().from(nivelesTable).where(eq(nivelesTable.id, reto.id_nivel));

    res.json({ ...reto, preguntas, modulo: modulo || null, nivel: nivel || null });
  } catch (err) {
    req.log.error({ err }, "Error al obtener reto");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /challenges/:challengeId/questions - Obtener solo las preguntas de un reto
router.get("/challenges/:challengeId/questions", requireAuth, async (req, res) => {
  try {
    const challengeId = Number(req.params.challengeId);
    const [reto] = await db.select().from(retosTable).where(eq(retosTable.id, challengeId));

    if (!reto) {
      res.status(404).json({ error: "not_found", message: "Reto no encontrado" });
      return;
    }

    // Shuffle aleatorio usando ORDER BY RANDOM() - diferente orden en cada intento
    const preguntas = await db
      .select()
      .from(preguntasTable)
      .where(
        and(
          eq(preguntasTable.id_modulo, reto.id_modulo),
          eq(preguntasTable.id_nivel, reto.id_nivel),
          eq(preguntasTable.activa, true)
        )
      )
      .orderBy(sql`RANDOM()`)
      .limit(reto.numero_preguntas);

    // Parsea las opciones que estan guardadas como JSON string
    const preguntasParseadas = preguntas.map((p) => ({
      ...p,
      opciones: typeof p.opciones === "string" ? JSON.parse(p.opciones) : p.opciones,
    }));

    res.json(preguntasParseadas);
  } catch (err) {
    req.log.error({ err }, "Error al obtener preguntas del reto");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// POST /challenges/:challengeId/check-answer - Verificar si una respuesta es correcta
router.post("/challenges/:challengeId/check-answer", requireAuth, async (req, res) => {
  try {
    const { pregunta_id, respuesta } = req.body;

    if (!pregunta_id || respuesta === undefined) {
      res.status(400).json({ error: "validation_error", message: "pregunta_id y respuesta son requeridos" });
      return;
    }

    const [pregunta] = await db
      .select()
      .from(preguntasTable)
      .where(eq(preguntasTable.id, Number(pregunta_id)));

    if (!pregunta) {
      res.status(404).json({ error: "not_found", message: "Pregunta no encontrada" });
      return;
    }

    const correcta =
      pregunta.respuesta_correcta.trim().toLowerCase() === String(respuesta).trim().toLowerCase();

    res.json({
      correcta,
      respuesta_correcta: pregunta.respuesta_correcta,
      explicacion: pregunta.explicacion || null,
      puntos: correcta ? pregunta.puntos : 0,
    });
  } catch (err) {
    req.log.error({ err }, "Error al verificar respuesta");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /challenges/custom - Obtener todos los retos personalizados
router.get("/challenges/custom", requireAuth, async (req, res) => {
  try {
    const retos = await db
      .select()
      .from(retosPersonalizadosTable)
      .where(eq(retosPersonalizadosTable.activo, true));

    const retosConInfo = await Promise.all(
      retos.map(async (reto) => {
        const [docente] = await db
          .select({ id: perfilesTable.id, nombre: perfilesTable.nombre, usuario: perfilesTable.usuario, rol: perfilesTable.rol })
          .from(perfilesTable)
          .where(eq(perfilesTable.id, reto.id_docente));
        const [modulo] = await db.select().from(modulosTable).where(eq(modulosTable.id, reto.id_modulo));
        const [nivel] = await db.select().from(nivelesTable).where(eq(nivelesTable.id, reto.id_nivel));
        return { ...reto, docente: docente || null, modulo: modulo || null, nivel: nivel || null };
      })
    );

    res.json(retosConInfo);
  } catch (err) {
    req.log.error({ err }, "Error al obtener retos personalizados");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// POST /challenges/custom - Crear un reto personalizado (solo docentes)
router.post("/challenges/custom", requireAuth, requireDocente, async (req: AuthRequest, res) => {
  try {
    const { nombre, descripcion, tipo_juego, id_modulo, id_nivel, puntos_maximos, numero_preguntas, tiempo_limite, preguntas } = req.body;

    if (!nombre || !tipo_juego || !id_modulo || !id_nivel) {
      res.status(400).json({ error: "validation_error", message: "Campos requeridos faltantes" });
      return;
    }

    // Crea el reto personalizado en la base de datos
    const [nuevoReto] = await db
      .insert(retosPersonalizadosTable)
      .values({
        id_docente: req.user!.id,
        nombre,
        descripcion,
        tipo_juego,
        id_modulo: Number(id_modulo),
        id_nivel: Number(id_nivel),
        puntos_maximos: puntos_maximos || 100,
        numero_preguntas: numero_preguntas || 10,
        tiempo_limite: tiempo_limite || 300,
      })
      .returning();

    // Si se proporcionaron preguntas, las guarda en la tabla de preguntas del reto
    if (preguntas && Array.isArray(preguntas) && preguntas.length > 0) {
      await db.insert(retosPersonalizadosPreguntasTable).values(
        preguntas.map((p: { texto: string; tipo: string; opciones?: string[]; respuesta_correcta: string; explicacion?: string; dificultad: string; puntos?: number; }) => ({
          id_reto: nuevoReto.id,
          id_pregunta: 0,
          texto: p.texto,
          tipo: p.tipo,
          opciones: p.opciones ? JSON.stringify(p.opciones) : null,
          respuesta_correcta: p.respuesta_correcta,
          explicacion: p.explicacion || null,
          dificultad: p.dificultad,
          puntos: p.puntos || 10,
        }))
      );
    }

    const [docente] = await db
      .select({ id: perfilesTable.id, nombre: perfilesTable.nombre, usuario: perfilesTable.usuario, rol: perfilesTable.rol })
      .from(perfilesTable)
      .where(eq(perfilesTable.id, req.user!.id));

    const [modulo] = await db.select().from(modulosTable).where(eq(modulosTable.id, nuevoReto.id_modulo));
    const [nivel] = await db.select().from(nivelesTable).where(eq(nivelesTable.id, nuevoReto.id_nivel));

    res.status(201).json({ ...nuevoReto, docente, modulo, nivel });
  } catch (err) {
    req.log.error({ err }, "Error al crear reto personalizado");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /challenges/custom/:challengeId - Obtener un reto personalizado con preguntas
router.get("/challenges/custom/:challengeId", requireAuth, async (req, res) => {
  try {
    const challengeId = Number(req.params.challengeId);
    const [reto] = await db
      .select()
      .from(retosPersonalizadosTable)
      .where(eq(retosPersonalizadosTable.id, challengeId));

    if (!reto) {
      res.status(404).json({ error: "not_found", message: "Reto personalizado no encontrado" });
      return;
    }

    // Obtiene las preguntas del reto personalizado
    const preguntasReto = await db
      .select()
      .from(retosPersonalizadosPreguntasTable)
      .where(eq(retosPersonalizadosPreguntasTable.id_reto, challengeId));

    const preguntas = preguntasReto.map((p) => ({
      ...p,
      opciones: p.opciones ? JSON.parse(p.opciones) : null,
    }));

    const [docente] = await db
      .select({ id: perfilesTable.id, nombre: perfilesTable.nombre, usuario: perfilesTable.usuario, rol: perfilesTable.rol })
      .from(perfilesTable)
      .where(eq(perfilesTable.id, reto.id_docente));
    const [modulo] = await db.select().from(modulosTable).where(eq(modulosTable.id, reto.id_modulo));
    const [nivel] = await db.select().from(nivelesTable).where(eq(nivelesTable.id, reto.id_nivel));

    res.json({ ...reto, preguntas, docente, modulo, nivel });
  } catch (err) {
    req.log.error({ err }, "Error al obtener reto personalizado");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// PUT /challenges/custom/:challengeId - Actualizar un reto personalizado
router.put("/challenges/custom/:challengeId", requireAuth, requireDocente, async (req: AuthRequest, res) => {
  try {
    const challengeId = Number(req.params.challengeId);
    const { nombre, descripcion, tipo_juego, id_modulo, id_nivel, puntos_maximos, numero_preguntas, tiempo_limite } = req.body;

    const [reto] = await db
      .select()
      .from(retosPersonalizadosTable)
      .where(eq(retosPersonalizadosTable.id, challengeId));

    if (!reto || reto.id_docente !== req.user!.id) {
      res.status(404).json({ error: "not_found", message: "Reto no encontrado o sin permiso" });
      return;
    }

    const [actualizado] = await db
      .update(retosPersonalizadosTable)
      .set({ nombre, descripcion, tipo_juego, id_modulo, id_nivel, puntos_maximos, numero_preguntas, tiempo_limite })
      .where(eq(retosPersonalizadosTable.id, challengeId))
      .returning();

    const [modulo] = await db.select().from(modulosTable).where(eq(modulosTable.id, actualizado.id_modulo));
    const [nivel] = await db.select().from(nivelesTable).where(eq(nivelesTable.id, actualizado.id_nivel));

    res.json({ ...actualizado, modulo, nivel });
  } catch (err) {
    req.log.error({ err }, "Error al actualizar reto personalizado");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// DELETE /challenges/custom/:challengeId - Eliminar un reto personalizado
router.delete("/challenges/custom/:challengeId", requireAuth, requireDocente, async (req: AuthRequest, res) => {
  try {
    const challengeId = Number(req.params.challengeId);

    const [reto] = await db
      .select()
      .from(retosPersonalizadosTable)
      .where(eq(retosPersonalizadosTable.id, challengeId));

    if (!reto || reto.id_docente !== req.user!.id) {
      res.status(404).json({ error: "not_found", message: "Reto no encontrado o sin permiso" });
      return;
    }

    // Elimina primero las preguntas del reto (foreign key)
    await db
      .delete(retosPersonalizadosPreguntasTable)
      .where(eq(retosPersonalizadosPreguntasTable.id_reto, challengeId));

    await db.delete(retosPersonalizadosTable).where(eq(retosPersonalizadosTable.id, challengeId));

    res.json({ success: true, message: "Reto eliminado exitosamente" });
  } catch (err) {
    req.log.error({ err }, "Error al eliminar reto personalizado");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

export default router;

File: artifacts/api-server-python/routers/challenges.py
import json
import random
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional, List, Any
from db import get_db, fetchone, fetchall, execute
from auth_utils import get_current_user, require_docente
from datetime import datetime

router = APIRouter()


def serialize_row(row):
    if row is None:
        return None
    r = dict(row)
    for k, v in r.items():
        if isinstance(v, datetime):
            r[k] = v.isoformat()
    return r


def parse_opciones(row: dict) -> dict:
    row = dict(row)
    if row.get("opciones") and isinstance(row["opciones"], str):
        try:
            row["opciones"] = json.loads(row["opciones"])
        except Exception:
            pass
    return row


@router.get("/challenges")
def get_challenges(id_modulo: int = None, user=Depends(get_current_user)):
    with get_db() as conn:
        if id_modulo:
            retos_predef = fetchall(conn, "SELECT * FROM retos WHERE activo = 1 AND id_modulo = %s", (id_modulo,))
        else:
            retos_predef = fetchall(conn, "SELECT * FROM retos WHERE activo = 1")

        predefinidos = []
        for r in retos_predef:
            r = dict(r)
            modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (r["id_modulo"],))
            nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (r["id_nivel"],))
            r["modulo"] = serialize_row(modulo)
            r["nivel"] = serialize_row(nivel)
            r["is_custom"] = False
            predefinidos.append(serialize_row(r))

        if id_modulo:
            custom_retos = fetchall(conn, "SELECT * FROM retos_personalizados WHERE activo = 1 AND publicado = 1 AND id_modulo = %s", (id_modulo,))
        else:
            custom_retos = fetchall(conn, "SELECT * FROM retos_personalizados WHERE activo = 1 AND publicado = 1")

        custom = []
        for r in custom_retos:
            r = dict(r)
            docente = fetchone(conn, "SELECT nombre FROM perfiles WHERE id = %s", (r["id_docente"],))
            modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (r["id_modulo"],))
            nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (r["id_nivel"],))
            r["id"] = -r["id"]
            r["modulo"] = serialize_row(modulo)
            r["nivel"] = serialize_row(nivel)
            r["is_custom"] = True
            r["docente_nombre"] = docente["nombre"] if docente else None
            custom.append(serialize_row(r))

    return predefinidos + custom


@router.get("/challenges/custom")
def get_custom_challenges(user=Depends(get_current_user)):
    with get_db() as conn:
        retos = fetchall(conn, "SELECT * FROM retos_personalizados WHERE activo = 1")
        result = []
        for r in retos:
            r = dict(r)
            docente = fetchone(conn, "SELECT id, nombre, usuario, rol FROM perfiles WHERE id = %s", (r["id_docente"],))
            modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (r["id_modulo"],))
            nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (r["id_nivel"],))
            r["docente"] = serialize_row(docente)
            r["modulo"] = serialize_row(modulo)
            r["nivel"] = serialize_row(nivel)
            result.append(serialize_row(r))
    return result


@router.get("/challenges/{challenge_id}")
def get_challenge(challenge_id: int, user=Depends(get_current_user)):
    with get_db() as conn:
        reto = fetchone(conn, "SELECT * FROM retos WHERE id = %s", (challenge_id,))
        if not reto:
            raise HTTPException(404, "Reto no encontrado")
        reto = dict(reto)
        preguntas = fetchall(conn,
            "SELECT * FROM preguntas WHERE id_modulo = %s AND id_nivel = %s AND activa = 1 LIMIT %s",
            (reto["id_modulo"], reto["id_nivel"], reto["numero_preguntas"]))
        preguntas = [parse_opciones(serialize_row(p)) for p in preguntas]
        modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (reto["id_modulo"],))
        nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (reto["id_nivel"],))
        reto["preguntas"] = preguntas
        reto["modulo"] = serialize_row(modulo)
        reto["nivel"] = serialize_row(nivel)
    return serialize_row(reto)


@router.get("/challenges/{challenge_id}/questions")
def get_challenge_questions(challenge_id: int, user=Depends(get_current_user)):
    with get_db() as conn:
        reto = fetchone(conn, "SELECT * FROM retos WHERE id = %s", (challenge_id,))
        if not reto:
            raise HTTPException(404, "Reto no encontrado")
        preguntas = fetchall(conn,
            "SELECT * FROM preguntas WHERE id_modulo = %s AND id_nivel = %s AND activa = 1 ORDER BY RAND() LIMIT %s",
            (reto["id_modulo"], reto["id_nivel"], reto["numero_preguntas"]))
    return [parse_opciones(serialize_row(p)) for p in preguntas]


class CheckAnswerBody(BaseModel):
    pregunta_id: int
    respuesta: Any


@router.post("/challenges/{challenge_id}/check-answer")
def check_answer(challenge_id: int, body: CheckAnswerBody, user=Depends(get_current_user)):
    with get_db() as conn:
        pregunta = fetchone(conn, "SELECT * FROM preguntas WHERE id = %s", (body.pregunta_id,))
    if not pregunta:
        raise HTTPException(404, "Pregunta no encontrada")
    correcta = pregunta["respuesta_correcta"].strip().lower() == str(body.respuesta).strip().lower()
    return {
        "correcta": correcta,
        "respuesta_correcta": pregunta["respuesta_correcta"],
        "explicacion": pregunta.get("explicacion"),
        "puntos": pregunta.get("puntos", 10) if correcta else 0,
    }


class PreguntaBody(BaseModel):
    texto: str
    tipo: str
    opciones: Optional[List[str]] = None
    respuesta_correcta: str
    explicacion: Optional[str] = None
    dificultad: str
    puntos: Optional[int] = 10


class CreateCustomBody(BaseModel):
    nombre: str
    descripcion: Optional[str] = None
    tipo_juego: str
    id_modulo: int
    id_nivel: int
    puntos_maximos: Optional[int] = 100
    numero_preguntas: Optional[int] = 10
    tiempo_limite: Optional[int] = 300
    preguntas: Optional[List[PreguntaBody]] = None


@router.post("/challenges/custom", status_code=201)
def create_custom_challenge(body: CreateCustomBody, user=Depends(get_current_user)):
    require_docente(user)
    if not body.nombre or not body.tipo_juego or not body.id_modulo or not body.id_nivel:
        raise HTTPException(400, "Campos requeridos faltantes")

    with get_db() as conn:
        with conn.cursor() as cur:
            cur.execute(
                """INSERT INTO retos_personalizados
                   (id_docente, nombre, descripcion, tipo_juego, id_modulo, id_nivel,
                    puntos_maximos, numero_preguntas, tiempo_limite, activo, publicado)
                   VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, 1, 0)""",
                (user["id"], body.nombre, body.descripcion, body.tipo_juego, body.id_modulo,
                 body.id_nivel, body.puntos_maximos, body.numero_preguntas, body.tiempo_limite),
            )
            conn.commit()
            nuevo_id = cur.lastrowid

        if body.preguntas:
            for p in body.preguntas:
                with conn.cursor() as cur:
                    cur.execute(
                        """INSERT INTO retos_personalizados_preguntas
                           (id_reto, id_pregunta, texto, tipo, opciones, respuesta_correcta, explicacion, dificultad, puntos)
                           VALUES (%s, 0, %s, %s, %s, %s, %s, %s, %s)""",
                        (nuevo_id, p.texto, p.tipo,
                         json.dumps(p.opciones) if p.opciones else None,
                         p.respuesta_correcta, p.explicacion, p.dificultad, p.puntos or 10),
                    )
            conn.commit()

        nuevo = fetchone(conn, "SELECT * FROM retos_personalizados WHERE id = %s", (nuevo_id,))
        docente = fetchone(conn, "SELECT id, nombre, usuario, rol FROM perfiles WHERE id = %s", (user["id"],))
        modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (body.id_modulo,))
        nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (body.id_nivel,))

    r = serialize_row(nuevo)
    r["docente"] = serialize_row(docente)
    r["modulo"] = serialize_row(modulo)
    r["nivel"] = serialize_row(nivel)
    return r


@router.get("/challenges/custom/{challenge_id}")
def get_custom_challenge(challenge_id: int, user=Depends(get_current_user)):
    with get_db() as conn:
        reto = fetchone(conn, "SELECT * FROM retos_personalizados WHERE id = %s", (challenge_id,))
        if not reto:
            raise HTTPException(404, "Reto personalizado no encontrado")
        reto = dict(reto)
        preguntas = fetchall(conn, "SELECT * FROM retos_personalizados_preguntas WHERE id_reto = %s", (challenge_id,))
        preguntas = [parse_opciones(serialize_row(p)) for p in preguntas]
        docente = fetchone(conn, "SELECT id, nombre, usuario, rol FROM perfiles WHERE id = %s", (reto["id_docente"],))
        modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (reto["id_modulo"],))
        nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (reto["id_nivel"],))
        reto["preguntas"] = preguntas
        reto["docente"] = serialize_row(docente)
        reto["modulo"] = serialize_row(modulo)
        reto["nivel"] = serialize_row(nivel)
    return serialize_row(reto)


class UpdateCustomBody(BaseModel):
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    tipo_juego: Optional[str] = None
    id_modulo: Optional[int] = None
    id_nivel: Optional[int] = None
    puntos_maximos: Optional[int] = None
    numero_preguntas: Optional[int] = None
    tiempo_limite: Optional[int] = None


@router.put("/challenges/custom/{challenge_id}")
def update_custom_challenge(challenge_id: int, body: UpdateCustomBody, user=Depends(get_current_user)):
    require_docente(user)
    with get_db() as conn:
        reto = fetchone(conn, "SELECT * FROM retos_personalizados WHERE id = %s", (challenge_id,))
        if not reto or reto["id_docente"] != user["id"]:
            raise HTTPException(404, "Reto no encontrado o sin permiso")

        updates = {k: v for k, v in body.dict().items() if v is not None}
        if updates:
            set_clause = ", ".join(f"{k} = %s" for k in updates)
            execute(conn, f"UPDATE retos_personalizados SET {set_clause} WHERE id = %s",
                    (*updates.values(), challenge_id))

        actualizado = fetchone(conn, "SELECT * FROM retos_personalizados WHERE id = %s", (challenge_id,))
        modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (actualizado["id_modulo"],))
        nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (actualizado["id_nivel"],))

    r = serialize_row(actualizado)
    r["modulo"] = serialize_row(modulo)
    r["nivel"] = serialize_row(nivel)
    return r


@router.patch("/challenges/custom/{challenge_id}/publish")
def toggle_publish(challenge_id: int, user=Depends(get_current_user)):
    require_docente(user)
    with get_db() as conn:
        reto = fetchone(conn, "SELECT * FROM retos_personalizados WHERE id = %s", (challenge_id,))
        if not reto or reto["id_docente"] != user["id"]:
            raise HTTPException(404, "Reto no encontrado o sin permiso")

        nuevo_estado = not bool(reto["publicado"])
        execute(conn, "UPDATE retos_personalizados SET publicado = %s WHERE id = %s",
                (1 if nuevo_estado else 0, challenge_id))

    return {
        "publicado": nuevo_estado,
        "message": "Reto publicado" if nuevo_estado else "Reto despublicado",
    }


@router.delete("/challenges/custom/{challenge_id}")
def delete_custom_challenge(challenge_id: int, user=Depends(get_current_user)):
    require_docente(user)
    with get_db() as conn:
        reto = fetchone(conn, "SELECT * FROM retos_personalizados WHERE id = %s", (challenge_id,))
        if not reto or reto["id_docente"] != user["id"]:
            raise HTTPException(404, "Reto no encontrado o sin permiso")
        execute(conn, "DELETE FROM retos_personalizados_preguntas WHERE id_reto = %s", (challenge_id,))
        execute(conn, "DELETE FROM retos_personalizados WHERE id = %s", (challenge_id,))
    return {"success": True, "message": "Reto eliminado exitosamente"}
